pkg/sources/registration.go

```diff
@@ -0,0 +1,195 @@
+				if len(sources) > 0 {
+					fn(ctx, resource, sources, diffs)(newObj)
+				}
+			},
+		})
+	}
+}
+
+func (i Registration) handleMapped(ctx context.Context, targetEvent config.EventType, routeTable map[string][]Entry, fn eventHandler) {
+	i.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			var eventObj coreV1.Event
+			err := utils.TransformIntoTypedObject(obj.(*unstructured.Unstructured), &eventObj)
+			if err != nil {
+				i.log.Errorf("Unable to transform object type: %v, into type: %v", reflect.TypeOf(obj), reflect.TypeOf(eventObj))
+			}
+			_, err = cache.MetaNamespaceKeyFunc(obj)
+			if err != nil {
+				i.log.Errorf("Failed to get MetaNamespaceKey from event resource")
+				return
+			}
+
+			// Find involved object type
+			gvr, err := utils.GetResourceFromKind(i.mapper, eventObj.InvolvedObject.GroupVersionKind())
+			if err != nil {
+				i.log.Errorf("Failed to get involved object: %v", err)
+				return
+			}
+
+			if !i.canHandleEvent(eventObj.Type) {
+				return
+			}
+
+			gvrToString := utils.GVRToString(gvr)
+			if !i.includesSrcResource(gvrToString) {
+				return
+			}
+
+			sourceRoutes := sourceRoutes(routeTable, gvrToString, targetEvent)
+			sources := sourcesForObjNamespace(ctx, sourceRoutes, obj, i.log, i.mapper, i.dynamicCli)
+			if len(sources) == 0 {
+				return
+			}
+			fn(ctx, gvrToString, sources, nil)(obj)
+		},
+	})
+}
+
+func (i Registration) canHandleEvent(target string) bool {
+	for _, e := range i.events {
+		if strings.ToLower(target) == e.String() {
+			return true
+		}
+	}
+	return false
+}
+
+func (i Registration) includesSrcResource(resource string) bool {
+	for _, src := range i.mappedResources {
+		if src == resource {
+			return true
+		}
+	}
+	return false
+}
+
+func sourcesForObjNamespace(ctx context.Context, routes []Route, obj interface{}, log logrus.FieldLogger, mapper meta.RESTMapper, cli dynamic.Interface) []string {
+	var out []string
+
+	objectMeta, err := utils.GetObjectMetaData(ctx, cli, mapper, obj)
+	if err != nil {
+		log.Errorf("while getting object metadata: %s", err.Error())
+		return []string{}
+	}
+
+	targetNs := objectMeta.Namespace
+	log.Debugf("handling events for target Namespace: %s in routes: %+v", targetNs, routes)
+
+	for _, route := range routes {
+		if route.namespaces.IsAllowed(targetNs) {
+			out = append(out, route.source)
+		}
+	}
+
+	return out
+}
+
+func qualifySourcesForUpdate(
+	ctx context.Context,
+	newObj, oldObj interface{},
+	routes []Route,
+	log logrus.FieldLogger,
+	mapper meta.RESTMapper,
+	cli dynamic.Interface,
+) ([]string, []string) {
+	var sources, diffs []string
+
+	candidates := sourcesForObjNamespace(ctx, routes, newObj, log, mapper, cli)
+
+	var oldUnstruct, newUnstruct *unstructured.Unstructured
+	var ok bool
+
+	if oldUnstruct, ok = oldObj.(*unstructured.Unstructured); !ok {
+		log.Error("Failed to typecast object to Unstructured.")
```

Comment:
what about `Failed to typecast old object to Unstructured.` to separate it from next message?